File: my-folder/0017-letter-combinations-of-a-phone-number/solution.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        keyboard = {"2": "abc",
                    "3": "def",
                    "4": "ghi",
                    "5": "jkl",
                    "6": "mno",
                    "7": "pqrs",
                    "8": "tuv",
                    "9": "wxyz"
                    }
        # for i in range(0,len(digits),2):
        #     list1=keyboard[digits[i]]
        #     list2=keyboard[digits[i+1]] if i+1<len(digits) else ''
        #     combs=[]
        #     for i in list1:
        #         for j in list2:
        #             combs.append(i+j)
        #     print(combs)
        
        res=[]
        path=[]
        def dfs(i):
            if(len(digits)==len(path)):
                res.append("".join(path))
                return 
            for digi in keyboard[digits[i]]:
                path.append(digi)
                dfs(i+1)
                path.pop()
        if digits:
            dfs(0)
        return res


            

                

           




        
        

File: my-folder/0017-letter-combinations-of-a-phone-number/solution.java
class Solution {
    Map<Integer, String> directory = new HashMap<>();;
    List<String> res = new ArrayList<>();;
    public List<String> letterCombinations(String digits) {
        if (digits == null || digits.length() == 0) return res;
        directory.put(2, "abc");
        directory.put(3, "def");
        directory.put(4, "ghi");
        directory.put(5, "jkl");
        directory.put(6, "mno");
        directory.put(7, "pqrs");
        directory.put(8, "tuv");
        directory.put(9, "wxyz");
 
        if(digits.length() == 1){
            int curr = digits.charAt(0) - '0';
            String letters = directory.get(curr);
            for(char letter : letters.toCharArray()){
                res.add(String.valueOf(letter));
            }
            return res;
        }
        
        dfs(0, "", digits);
        return res;
    }

    private void dfs(int idx, String path, String digits){
        if(path.length() == digits.length()){
            res.add(path);
            return;
        }

        int curr = digits.charAt(idx) - '0';
        String letters = directory.get(curr);
        for(char letter : letters.toCharArray()){
            dfs(idx + 1, path + letter, digits);
        }

    }
}

// 2 - 9 digits 
// digits = 2323

// ab ae af bd be bf ce cf cd

// do dfs on all the digits

